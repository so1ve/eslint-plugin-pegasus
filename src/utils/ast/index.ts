export * from "./call-or-new-expression";
export * from "./is-member-expression";
export * from "./is-method-call";
