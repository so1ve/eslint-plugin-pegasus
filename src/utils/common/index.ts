export * from "./is-node-matches";
export * from "./is-number";
export * from "./parentheses";
export * from "./rule";
