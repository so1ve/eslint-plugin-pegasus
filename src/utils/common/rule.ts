import type { TSESLint } from "@typescript-eslint/utils";

const isIterable = (object: any): object is Iterable<any> =>
	typeof object?.[Symbol.iterator] === "function";

class FixAbortError extends Error {}
const fixOptions = {
	abort() {
		throw new FixAbortError("Fix aborted.");
	},
};

export type AbortFunction = () => void;

type WrappedFixFunction = (
	fixer: TSESLint.RuleFixer,
	additional: { abort: AbortFunction },
) =>
	| IterableIterator<TSESLint.RuleFix>
	| TSESLint.RuleFix
	| readonly TSESLint.RuleFix[]
	| null;

export const wrapFixFunction =
	(fix: WrappedFixFunction): TSESLint.ReportFixFunction =>
	(fixer) => {
		const result = fix(fixer, fixOptions);

		if (isIterable(result)) {
			try {
				return [...result];
			} catch (error) {
				if (error instanceof FixAbortError) {
					return [];
				}

				throw error;
			}
		}

		return result;
	};
