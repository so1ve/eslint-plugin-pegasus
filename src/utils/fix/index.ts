export * from "./remove-member-expression-property";
export * from "./remove-method-call";
export * from "./replace-argument";
