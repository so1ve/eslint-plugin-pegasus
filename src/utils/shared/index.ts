export * from "./simple-array-search-rule";
