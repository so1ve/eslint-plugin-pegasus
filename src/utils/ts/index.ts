import typeutils from "@typescript-eslint/type-utils";
import type {
	ParserServicesWithTypeInformation,
	TSESTree,
} from "@typescript-eslint/typescript-estree";
import * as tsutils from "ts-api-utils";
import ts from "typescript";

export class TypeHelper {
	typeChecker: ts.TypeChecker;

	constructor(public services: ParserServicesWithTypeInformation) {
		this.typeChecker = services.program.getTypeChecker();
	}

	private getType(node: TSESTree.Node) {
		return typeutils.getConstrainedTypeAtLocation(this.services, node);
	}

	isTypeArrayTypeOrUnionOfArrayTypes(node: TSESTree.Node) {
		const nodeType = this.getType(node);

		return typeutils.isTypeArrayTypeOrUnionOfArrayTypes(
			nodeType,
			this.typeChecker,
		);
	}

	isNumber(node: TSESTree.Node) {
		const nodeType = this.getType(node);

		return tsutils.isTypeFlagSet(nodeType, ts.TypeFlags.Number);
	}

	isAny(node: TSESTree.Node) {
		const nodeType = this.getType(node);

		return typeutils.isTypeAnyType(nodeType);
	}
}
