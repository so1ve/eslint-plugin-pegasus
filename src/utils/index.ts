import { ESLintUtils } from "@typescript-eslint/utils";

export const createRule = ESLintUtils.RuleCreator(
	(ruleName) =>
		`https://github.com/so1ve/eslint-plugin-pegasus/blob/main/docs/rules/${ruleName}.md`,
);
