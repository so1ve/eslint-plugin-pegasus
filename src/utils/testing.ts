import { resolve } from "node:path";

import type { TSESLint } from "@typescript-eslint/utils";
import { expect } from "vitest";

import { RuleTester } from "../../vendor/rule-tester/src/RuleTester";

const fixturesDir = resolve(process.cwd(), "test", "fixtures");

export function getRuleTester() {
	const ruleTester = new RuleTester({
		parser: require.resolve("@typescript-eslint/parser"),
		parserOptions: {
			ecmaVersion: 2018,
			tsconfigRootDir: fixturesDir,
			project: "./tsconfig.json",
		},
	});

	return ruleTester;
}

export function test<
	MessageIds extends string,
	Options extends readonly unknown[] = [],
	RuleListener extends TSESLint.RuleListener = TSESLint.RuleListener,
>(
	ruleName: string,
	rule: TSESLint.RuleModule<MessageIds, Options, RuleListener>,
	{
		valid,
		invalid,
	}: {
		valid: string[];
		invalid: TSESLint.InvalidTestCase<MessageIds, Options>[];
	},
) {
	const ruleTester = getRuleTester();

	ruleTester.run(ruleName, rule, {
		valid,
		invalid,
	});
}

export function testSnapshot<
	MessageIds extends string,
	Options extends readonly unknown[] = [],
	RuleListener extends TSESLint.RuleListener = TSESLint.RuleListener,
>(
	ruleName: string,
	rule: TSESLint.RuleModule<MessageIds, Options, RuleListener>,
	{
		valid = [],
		invalid = [],
	}: {
		valid?: string[];
		invalid?: string[];
	},
) {
	const ruleTester = getRuleTester();

	ruleTester.run(ruleName, rule, {
		valid,
		invalid: invalid.map((i) => ({
			code: i,
			errors: null,
			onOutput: (output: string) => {
				expect(output).toMatchSnapshot();
			},
		})),
	});
}
