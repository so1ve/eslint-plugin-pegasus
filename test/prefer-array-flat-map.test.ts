import dedent from "dedent";

import rule, { RULE_NAME } from "../src/rules/prefer-array-flat-map";
import { testSnapshot } from "../src/utils/testing";

describe("prefer-array-flat-map", () => {
	testSnapshot(RULE_NAME, rule, {
		valid: [
			"const bar = [1,2,3].map()",
			"const bar = [1,2,3].map(i => i)",
			"const bar = [1,2,3].map((i) => i)",
			"const bar = [1,2,3].map((i) => { return i; })",
			"const bar = foo.map(i => i)",
			"const bar = [[1],[2],[3]].flat()",
			"const bar = [1,2,3].map(i => [i]).sort().flat()",
			dedent`
				let bar = [1,2,3].map(i => [i]);
				bar = bar.flat();
			`,
			"const bar = [[1],[2],[3]].map(i => [i]).flat(2)",
			"const bar = [[1],[2],[3]].map(i => [i]).flat(1, null)",
			"const bar = [[1],[2],[3]].map(i => [i]).flat(Infinity)",
			"const bar = [[1],[2],[3]].map(i => [i]).flat(Number.POSITIVE_INFINITY)",
			"const bar = [[1],[2],[3]].map(i => [i]).flat(Number.MAX_VALUE)",
			"const bar = [[1],[2],[3]].map(i => [i]).flat(Number.MAX_SAFE_INTEGER)",
			"const bar = [[1],[2],[3]].map(i => [i]).flat(...[1])",
			"const bar = [[1],[2],[3]].map(i => [i]).flat(0.4 +.6)",
			"const bar = [[1],[2],[3]].map(i => [i]).flat(+1)",
			"const bar = [[1],[2],[3]].map(i => [i]).flat(foo)",
			"const bar = [[1],[2],[3]].map(i => [i]).flat(foo.bar)",
			"const bar = [[1],[2],[3]].map(i => [i]).flat(1.00)",
			"const bar = { map: () => {} }.map(i => [i]).flat()", // not an array
		],
		invalid: [
			"const bar = [[1],[2],[3]].map(i => [i]).flat()",
			"const bar = [[1],[2],[3]].map(i => [i]).flat(1,)",
			"const bar = [1,2,3].map(i => [i]).flat()",
			"const bar = [1,2,3].map((i) => [i]).flat()",
			"const bar = [1,2,3].map((i) => { return [i]; }).flat()",
			"const bar = [1,2,3].map(foo).flat()",
			"const bar = [1,2,3].map(i => i).map(i => [i]).flat()",
			"const bar = [1,2,3].sort().map(i => [i]).flat()",
			"const bar = (([1,2,3].map(i => [i]))).flat()",
			dedent`
			let bar = [1,2,3].map(i => {
				return [i];
			}).flat();
		`,
			dedent`
			let bar = [1,2,3].map(i => {
				return [i];
			})
			.flat();
		`,
			dedent`
			let bar = [1,2,3].map(i => {
				return [i];
			}) // comment
			.flat();
		`,
			dedent`
			let bar = [1,2,3].map(i => {
				return [i];
			}) // comment
			.flat(); // other
		`,
			dedent`
			let bar = [1,2,3]
				.map(i => { return [i]; })
				.flat();
		`,
			dedent`
			let bar = [1,2,3].map(i => { return [i]; })
				.flat();
		`,
			"let bar = [1,2,3] . map( x => y ) . flat () // ðŸ¤ª",
			"const bar = [1,2,3].map(i => [i]).flat(1);",
		],
	});
});
