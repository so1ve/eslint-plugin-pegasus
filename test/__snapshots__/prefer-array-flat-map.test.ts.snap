// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > const bar = (([1,2,3].map(i => [i]))).flat() 1`] = `"const bar = (([1,2,3].flatMap(i => [i])))"`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > const bar = [[1],[2],[3]].map(i => [i]).flat() 1`] = `"const bar = [[1],[2],[3]].flatMap(i => [i])"`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > const bar = [[1],[2],[3]].map(i => [i]).flat(1,) 1`] = `"const bar = [[1],[2],[3]].flatMap(i => [i])"`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > const bar = [1,2,3].map((i) => [i]).flat() 1`] = `"const bar = [1,2,3].flatMap((i) => [i])"`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > const bar = [1,2,3].map((i) => { return [i]; }).flat() 1`] = `"const bar = [1,2,3].flatMap((i) => { return [i]; })"`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > const bar = [1,2,3].map(foo).flat() 1`] = `"const bar = [1,2,3].flatMap(foo)"`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > const bar = [1,2,3].map(i => [i]).flat() 1`] = `"const bar = [1,2,3].flatMap(i => [i])"`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > const bar = [1,2,3].map(i => [i]).flat(1); 1`] = `"const bar = [1,2,3].flatMap(i => [i]);"`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > const bar = [1,2,3].map(i => i).map(i => [i]).flat() 1`] = `"const bar = [1,2,3].map(i => i).flatMap(i => [i])"`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > const bar = [1,2,3].sort().map(i => [i]).flat() 1`] = `"const bar = [1,2,3].sort().flatMap(i => [i])"`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > let bar = [1,2,3]
\\u0009.map(i => { return [i]; })
\\u0009.flat(); 1`] = `
"let bar = [1,2,3]
	.flatMap(i => { return [i]; });"
`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > let bar = [1,2,3] . map( x => y ) . flat () // ğŸ¤ª 1`] = `"let bar = [1,2,3] . flatMap( x => y ) // ğŸ¤ª"`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > let bar = [1,2,3].map(i => {
\\u0009return [i];
})
.flat(); 1`] = `
"let bar = [1,2,3].flatMap(i => {
	return [i];
});"
`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > let bar = [1,2,3].map(i => {
\\u0009return [i];
}) // comment
.flat(); // other 1`] = `
"let bar = [1,2,3].flatMap(i => {
	return [i];
}); // other"
`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > let bar = [1,2,3].map(i => {
\\u0009return [i];
}) // comment
.flat(); 1`] = `
"let bar = [1,2,3].flatMap(i => {
	return [i];
});"
`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > let bar = [1,2,3].map(i => {
\\u0009return [i];
}).flat(); 1`] = `
"let bar = [1,2,3].flatMap(i => {
	return [i];
});"
`;

exports[`prefer-array-flat-map > prefer-array-flat-map > invalid > let bar = [1,2,3].map(i => { return [i]; })
\\u0009.flat(); 1`] = `"let bar = [1,2,3].flatMap(i => { return [i]; });"`;
